training/unit-testing/go-lang/README.md

```diff
@@ -0,0 +1,126 @@
+- [Downloadable order of execution example](#orderexecution/execution_order_example_test.go)
+- [GoConvey explanation](https://github.com/smartystreets/goconvey/wiki/Execution-order)
+
+#### Assertions
+- When comparing objects in your tests you can use [GoConvey's Assertions](/assertions/README.md). 
+
+### Creating Easy To Test Code
+#### Writing SOLID Go
+Code that is easy to test is often code which follows the [SOLID](https://github.com/iamharvey/SOLID_principles) principles.  Here is good video on writing [SOLID Go](https://www.youtube.com/watch?v=zzAdEt3xZ1M).
+
+#### Dependency Injection
+This allows objects to be more independent of their main dependencies, for example we could create a website templating module which could implement all the functionality required to template the website, minus the rendering.  This rendering functionality could be injected, for example for devices such as mobile, tablet, laptop.  The rendering on the different browsers could then be injected into the rendering routines for the different devices... and so the design unfolds.
+
+The following [walk-through in GoLang is useful](https://quii.gitbook.io/learn-go-with-tests/go-fundamentals/dependency-injection)
+
+Another occasion when Dependency Injection is useful is when making mockups.  The following downloadable example walks through this technique:
+- [Dependency Injection within Mockups](https://github.com/sohamkamani/go-dependency-injection-example)
+
+#### Test Driven Development
+A useful approach, particularly for potentially difficult to test code, is Test Driven Development, which involves writing the test at the same time or before the code you need to test. So effectively designing easy to test code from the start.
+- [Golang Test Driven Development walk through](https://itnext.io/how-to-tdd-a-console-application-to-achieve-100-coverage-starting-from-main-test-go-934a617b080f) 
+
+### Recommended Go Tutorials
+- [GoConvey Tutorial 1](https://www.smartystreets.com/blog/2015/02/go-testing-part-1-vanillla/)
+- [GoConvey Tutorial 2](https://www.smartystreets.com/blog/2015/02/go-testing-part-2-running-tests/)
+- [GoConvey Tutorial 3](https://www.smartystreets.com/blog/2015/05/go-testing-part-3-convey-behavior/)
+
+### Code Coverage
+When we test certain paths of execution are tested, but often not all.  Knowing which paths are tested is useful, but often the same paths need to be tested multiple time, so 100% coverage doesn't mean 100% tested.
+- [Excellent Golang code coverage walk-through](https://blog.golang.org/cover)
+- [Useful time saving suggestion on stackoverflow](https://stackoverflow.com/a/27284510)
+
+### Assessing How Fast Your Code Runs
+Systems often need to run efficiently, such as a response from a website.  Go incorporates a benchmarking routines, which will time how long the code will take to run.  It clever though as it repeats the tests until it feel it's got an accurate time.  
```

Comment:
Again extra whitespace between `a website. ` and ` Go incorporates` and then `run. ` and ` It clever` - Im now wondering if this wasn't by accident?

Reply:
See comment above.